**Emmanuel Decena - ME EE**
Deep Learning Assignment

Build a classifier on CIFAR10 dataset using
- 3-layer MLP
- 3-layer CNN

Compare performance

**1. Using 3-Layer MLP**

In [1]:
import tensorflow as tf
import numpy as np

from keras.datasets import cifar10
from keras.utils import np_utils
from keras.models import Sequential
from keras.layers.core import Dense, Flatten, Activation
from keras.layers import Conv2D, MaxPooling2D
from keras.callbacks import ModelCheckpoint
from keras.preprocessing.image import ImageDataGenerator
from keras.utils import to_categorical, plot_model

In [6]:
# load dataset
(x_train, y_train), (x_test, y_test) = cifar10.load_data()
num_labels = len(np.unique(y_train))
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

image_size = x_train.shape[1]
input_size = image_size * image_size * 3

# resize and normalize
x_train = np.reshape(x_train, [-1, input_size]) 
x_train = x_train.astype('float32') / 255
x_test = np.reshape(x_test, [-1, input_size])
x_test = x_test.astype('float32') / 255

In [7]:
# hyper-parameters
batch_size = 256
n_units = 256
epochs = 10

In [8]:
# model is a 3-layer MLP
model = Sequential(name='3-layer-MLP')
model.add(Dense(n_units, input_dim=input_size, activation='relu', name='input_layer'))
model.add(Dense(n_units, activation='relu', name='hidden_layer'))
model.add(Dense(num_labels, activation='softmax', name='output_layer'))
model.summary()

Model: "3-layer-MLP"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_layer (Dense)          (None, 256)               786688    
_________________________________________________________________
hidden_layer (Dense)         (None, 256)               65792     
_________________________________________________________________
output_layer (Dense)         (None, 10)                2570      
Total params: 855,050
Trainable params: 855,050
Non-trainable params: 0
_________________________________________________________________


In [9]:
model.compile(loss="categorical_crossentropy", optimizer="sgd", metrics=["accuracy"])

history = model.fit(
    x_train,
    y_train,
    validation_data=(x_test, y_test),
    epochs=epochs,
    batch_size=batch_size,
)

loss, acc = model.evaluate(x_test, y_test, batch_size=batch_size, verbose=False)

print("\nTest accuracy: %.1f%%" % (100.0 * acc))


Epoch 1/10
196/196 [==============================] - 4s 21ms/step - loss: 2.0759 - accuracy: 0.2589 - val_loss: 1.9620 - val_accuracy: 0.2861
Epoch 2/10
196/196 [==============================] - 4s 19ms/step - loss: 1.8968 - accuracy: 0.3345 - val_loss: 1.8576 - val_accuracy: 0.3447
Epoch 3/10
196/196 [==============================] - 3s 17ms/step - loss: 1.8329 - accuracy: 0.3574 - val_loss: 1.8351 - val_accuracy: 0.3565
Epoch 4/10
196/196 [==============================] - 4s 20ms/step - loss: 1.7909 - accuracy: 0.3745 - val_loss: 1.7859 - val_accuracy: 0.3715
Epoch 5/10
196/196 [==============================] - 4s 22ms/step - loss: 1.7568 - accuracy: 0.3865 - val_loss: 1.7802 - val_accuracy: 0.3757
Epoch 6/10
196/196 [==============================] - 4s 21ms/step - loss: 1.7273 - accuracy: 0.3949 - val_loss: 1.7253 - val_accuracy: 0.3979
Epoch 7/10
196/196 [==============================] - 4s 20ms/step - loss: 1.7048 - accuracy: 0.4029 - val_loss: 1.7047 - val_accuracy: 0.4037

**2. Using 3-Layer CNN**

In [10]:
# load mnist dataset
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

In [11]:
# input image dimensions
image_size = x_train.shape[1]
# resize and normalize
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)
x_train = x_train.astype('float32') / 255
x_test = x_test.astype('float32') / 255

# network parameters
input_shape = (image_size, image_size, 3)
batch_size = 256
kernel_size = 3
pool_size = 3
filters = 32

In [12]:
model = Sequential()
model.add(Conv2D(filters=filters,
                 kernel_size=kernel_size, 
                 activation='relu', 
                 kernel_initializer='he_uniform', 
                 padding='same', 
                 input_shape=input_shape,
                 name='input_layer'))

model.add(MaxPooling2D(pool_size,name='hidden_layer'))
model.add(Flatten())
model.add(Dense(10, activation='softmax',name='output_layer'))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_layer (Conv2D)         (None, 32, 32, 32)        896       
_________________________________________________________________
hidden_layer (MaxPooling2D)  (None, 10, 10, 32)        0         
_________________________________________________________________
flatten (Flatten)            (None, 3200)              0         
_________________________________________________________________
output_layer (Dense)         (None, 10)                32010     
Total params: 32,906
Trainable params: 32,906
Non-trainable params: 0
_________________________________________________________________


In [14]:
model.compile(loss="categorical_crossentropy", optimizer="sgd", metrics=["accuracy"])
# train the network
model.fit(x_train, y_train, epochs=epochs, batch_size=batch_size)

loss, acc = model.evaluate(x_test, y_test, batch_size=batch_size)
print("\nTest accuracy: %.1f%%" % (100.0 * acc))

Epoch 1/10
196/196 [==============================] - 21s 106ms/step - loss: 1.0762 - accuracy: 0.6346
Epoch 2/10
196/196 [==============================] - 18s 93ms/step - loss: 1.0605 - accuracy: 0.6400
Epoch 3/10
196/196 [==============================] - 20s 104ms/step - loss: 1.0471 - accuracy: 0.6424
Epoch 4/10
196/196 [==============================] - 22s 115ms/step - loss: 1.0493 - accuracy: 0.6435
Epoch 5/10
196/196 [==============================] - 21s 109ms/step - loss: 1.0413 - accuracy: 0.6470
Epoch 6/10
196/196 [==============================] - 22s 112ms/step - loss: 1.0417 - accuracy: 0.6441
Epoch 7/10
196/196 [==============================] - 24s 122ms/step - loss: 1.0336 - accuracy: 0.6497
Epoch 8/10
196/196 [==============================] - 20s 102ms/step - loss: 1.0328 - accuracy: 0.6493
Epoch 9/10
196/196 [==============================] - 20s 102ms/step - loss: 1.0344 - accuracy: 0.6479
Epoch 10/10
40/40 [==============================] - 1s 33ms/step - loss: 

**3. Performance**

The MLP model was outperformed by the CNN model using the same epochs and layers. 
The MLP's accuracy of ~30% was lower compared to the CNN accuracy of ~50%.
The MLP also used approximately ten times as much parameters at 855,050 compared to the CNN at 82,346.
The CNN model was more parameter efficient and has higher accuracy and thus is more suitable to use for this dataset.